quickwit/quickwit-actors/src/supervisor.rs

```diff
@@ -0,0 +1,383 @@
+        format!("Supervisor({})", self.actor_name)
+    }
+
+    fn queue_capacity(&self) -> crate::QueueCapacity {
+        crate::QueueCapacity::Unbounded
+    }
+
+    async fn initialize(&mut self, ctx: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+        ctx.schedule_self_msg(crate::HEARTBEAT, SuperviseLoop).await;
+        Ok(())
+    }
+
+    async fn finalize(
+        &mut self,
+        exit_status: &ActorExitStatus,
+        _ctx: &ActorContext<Self>,
+    ) -> anyhow::Result<()> {
+        match exit_status {
+            ActorExitStatus::Quit => {
+                if let Some(handle) = self.handle_opt.take() {
+                    handle.quit().await;
+                }
+            }
+            ActorExitStatus::Killed => {
+                if let Some(handle) = self.handle_opt.take() {
+                    handle.activate_kill_switch();
+                }
+            }
+            ActorExitStatus::Failure(_)
+            | ActorExitStatus::Success
+            | ActorExitStatus::DownstreamClosed => {}
+            ActorExitStatus::Panicked => {}
+        }
+
+        Ok(())
+    }
+}
+
+impl<A: Actor> Supervisor<A> {
+    async fn supervise(
+        &mut self,
+        ctx: &ActorContext<Supervisor<A>>,
+    ) -> Result<(), ActorExitStatus> {
+        match self.handle_opt.as_ref().unwrap().health() {
+            Health::Healthy => {
+                return Ok(());
+            }
+            Health::FailureOrUnhealthy => {}
+            Health::Success => {
+                return Err(ActorExitStatus::Success);
+            }
+        }
+        warn!("unhealthy-actor");
+        // The actor is failing we need to restart it.
+        let actor_handle = self.handle_opt.take().unwrap();
+        let (actor_exit_status, _last_state) = if actor_handle.state() == ActorState::Processing {
+            // The actor is probably frozen.
+            // Let's kill it.
+            warn!("killing");
+            actor_handle.kill().await
+        } else {
+            actor_handle.join().await
+        };
+        match actor_exit_status {
+            ActorExitStatus::Success => {
+                return Err(ActorExitStatus::Success);
+            }
+            ActorExitStatus::Quit => {
+                return Err(ActorExitStatus::Quit);
+            }
+            ActorExitStatus::DownstreamClosed => {
+                return Err(ActorExitStatus::DownstreamClosed);
+            }
+            ActorExitStatus::Killed => {
+                self.state.num_kills += 1;
+            }
+            ActorExitStatus::Failure(_err) => {
+                self.state.num_errors += 1;
+            }
+            ActorExitStatus::Panicked => {
+                self.state.num_panics += 1;
+            }
+        }
+        info!("respawning-actor");
+        let (_, actor_handle) = ctx
+            .spawn_actor()
+            .set_mailboxes(self.mailbox.clone(), self.inbox.clone())
+            .set_kill_switch(KillSwitch::default())
```

Comment:
We probably want to use a supervisor child kill switch

Reply:
oups yes absolutely.